File: adventofcode.tests.csharp/2020/2020_12.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace Year2020
{
    public static class AOC2020_12
    {
        private const int North = 0;
        private const int East = 1;
        private const int South = 2;
        private const int West = 3;

        static long CalcDist(int[] pos)
        {
            int x = Math.Abs(pos[East] - pos[West]);
            int y = Math.Abs(pos[South] - pos[North]);

            return x + y;
        }

        static int Forward(int[] pos, int[] dirs, int amount)
        {
            for (int i = 0; i < 4; i++)
            {
                pos[i] += dirs[i] * amount;
            }

            return 0;
        }

        static void Rotate(int[] dirs, int amount)
        {
            while (amount < 0)
            {
                int temp = dirs[3];
                dirs[3] = dirs[0];
                dirs[0] = dirs[1];
                dirs[1] = dirs[2];
                dirs[2] = temp;

                amount += 90;
            }

            while (amount > 0)
            {
                int temp = dirs[3];
                dirs[3] = dirs[2];
                dirs[2] = dirs[1];
                dirs[1] = dirs[0];
                dirs[0] = temp;

                amount -= 90;
            }
        }

        public static Tuple<long, long> Solve(string[] lines)
        {
            List<(char, int)> instructions = lines.Select(x => (x[0], int.Parse(x[1..]))).ToList();

            // N | E | S | W
            int[] pos1 = new[] { 0, 0, 0, 0 };
            int[] dirs1 = new[] { 0, 1, 0, 0 };

            foreach ((char, int) inst in instructions)
            {
                switch (inst.Item1)
                {
                    case 'N': { pos1[North] += inst.Item2; break; }
                    case 'E': { pos1[East] += inst.Item2; break; }
                    case 'S': { pos1[South] += inst.Item2; break; }
                    case 'W': { pos1[West] += inst.Item2; break; }
                    case 'L': { Rotate(dirs1, -1 * inst.Item2); break; }
                    case 'R': { Rotate(dirs1, inst.Item2); break; }
                    case 'F': { Forward(pos1, dirs1, inst.Item2); break; }
                };
            }

            // N | E | S | W
            int[] pos2 = new[] { 0, 0, 0, 0 };
            int[] wpRel = new[] { 1, 10, 0, 0 };

            foreach ((char, int) inst in instructions)
            {
                switch (inst.Item1)
                {
                    case 'N': { wpRel[North] += inst.Item2; break; }
                    case 'E': { wpRel[East] += inst.Item2; break; }
                    case 'S': { wpRel[South] += inst.Item2; break; }
                    case 'W': { wpRel[West] += inst.Item2; break; }
                    case 'L': { Rotate(wpRel, -1 * inst.Item2); break; }
                    case 'R': { Rotate(wpRel, inst.Item2); break; }
                    case 'F': { Forward(pos2, wpRel, inst.Item2); break; }
                };
            }

            long ans1 = CalcDist(pos1);
            long ans2 = CalcDist(pos2);

            return Tuple.Create(ans1, ans2);
        }
    }

    public class Day12
    {
        [Fact]
        public void Sample()
        {
            var lines = Utils.ReadInput("2020", "12", "sample");
            Utils.SolveAndValidate(Tuple.Create(25L, 286L), AOC2020_12.Solve, lines);
        }

        [Fact]
        public void Actual()
        {
            var lines = Utils.ReadInput("2020", "12", "actual");
            Utils.SolveAndValidate(Tuple.Create(998L, 71586L), AOC2020_12.Solve, lines);
        }
    }
}

File: adventofcode.tests.fsharp/2020/2020_12.fs
namespace Year2020

module Day12 =
    open Utils
    open Xunit

    [<Literal>]
    let North = 0

    [<Literal>]
    let East = 1

    [<Literal>]
    let South = 2

    [<Literal>]
    let West = 3

    type Instruction =
        { Direction: char;
          Amount: int }

    let dirIdx direction =
        match direction with
        | 'N' -> North
        | 'E' -> East
        | 'S' -> South
        | 'W' -> West
        | _ -> failwith $"Unsupported direction {direction}"

    let rotate amount arr =
        let rotateLeft (a : int array) =
            let tmp = a.[0]
            a.[0] <- a.[1]
            a.[1] <- a.[2]
            a.[2] <- a.[3]
            a.[3] <- tmp

        let rotateRight (a : int array) =
            let tmp = a.[3]
            a.[3] <- a.[2]
            a.[2] <- a.[1]
            a.[1] <- a.[0]
            a.[0] <- tmp

        let numMoves = amount / 90 |> abs
        let func = if amount < 0 then rotateLeft else rotateRight
        [1..numMoves] |> List.iter (fun _ -> func arr)

    let forward amount (dirs : int array) (pos : int array) =
        for i in 0..3 do
            pos.[i] <- (pos.[i] + (dirs.[i] * amount))

    let move amount dir (pos : int array) =
        let idx = dirIdx dir
        pos.[idx] <- pos.[idx] + amount

    let calcDistance (positions : int array) =
        let x = positions.[East] - positions.[West] |> abs
        let y = positions.[North] - positions.[South] |> abs
        x + y |> int64

    let parse (lines : string[]) =
        lines |> Array.map (fun line -> { Direction = line.[0]; Amount = line.[1..] |> int })

    let solve1 instructions =
        let mutable pos = [|0; 0; 0; 0|]
        let mutable dirs = [|0; 1; 0; 0|]

        for inst in instructions do
            match inst.Direction with
            | 'N' | 'E' | 'S' | 'W' -> move inst.Amount inst.Direction pos
            | 'L' -> rotate (inst.Amount * -1) dirs
            | 'R' -> rotate inst.Amount dirs
            | 'F' -> forward inst.Amount dirs pos
            | _ -> failwith $"Unknown instruction '{inst.Direction}'"
            |> ignore

        pos |> calcDistance |> int64

    let solve2 instructions =
        let mutable pos = [|0; 0; 0; 0|]
        let mutable wp = [|1; 10; 0; 0|]

        for inst in instructions do
            match inst.Direction with
            | 'N' | 'E' | 'S' | 'W' -> move inst.Amount inst.Direction wp
            | 'L' -> rotate (inst.Amount * -1) wp
            | 'R' -> rotate inst.Amount wp
            | 'F' -> forward inst.Amount wp pos
            | _ -> failwith $"Unknown instruction '{inst.Direction}'"
            |> ignore

        pos |> calcDistance |> int64

    let solve (lines : string[]) =
        let instructions = parse lines

        let ans1 = instructions |> solve1
        let ans2 = instructions |> solve2

        (ans1, ans2)

    [<Fact>]
    let Sample () =
        readInput "2020" "12" "sample" |> solveAndValidate (25L, 286L) solve

    [<Fact>]
    let Actual () =
        readInput "2020" "12" "actual" |> solveAndValidate (998L, 71586L) solve